### DISCUSSION TITLE
Help disabling default `K` keybind from LSP

### DISCUSSION THREAD

**Discussion by heysokam** [LazyVim/LazyVim#3880] [Q&A] [answered]
First of all, I have read https://www.lazyvim.org/plugins/lsp 4 different times
I have also successfully configured this keybind in a raw neovim+LSP setup before.
This is a question specific to LazyVim, as I cannot figure out how to disable the `K` keybind from the lsp config no matter what I try

This is my current setup in `config/options.lua`, in case there is some typo or something inherently wrong about it:
```lua
-- Options are automatically loaded before lazy.nvim startup
-- Default options that are always set: https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/config/options.lua
-- Add any additional options here

vim.g.autoformat     = false
vim.g.lazygit_config = false

return {
  "neovim/nvim-lspconfig",
  init = function()
    local keys = require("lazyvim.plugins.lsp.keymaps").get()
    -- change a keymap
    --keys[#keys + 1] = { "K", "<cmd>echo 'hello'<cr>" }
    -- add a keymap
    --keys[#keys + 1] = { "H", "<cmd>echo 'hello'<cr>" }
    -- disable a keymap
    keys[#keys + 1] = { "K", false }
  end,
  opts = {
    autoformat = false,
    servers = {
      tsserver = {
        autostart = false,
        init_options = {
          preferences = {
            importModuleSpecifierPreference = "relative",
            importModuleSpecifierEnding = "minimal",
          },
        },
      },
    },
  },
}
```

What am I doing wrong? :thinking:
Is this not what the very docs say I should do?

**Comment by Old-Farmer** [comment]
In neovim 0.10.0，K in Normal mode maps to vim.lsp.buf.hover() by default when lsp client starts, so deleting the key in this way  can not take effects. See `help lsp-defaults`

**Reply by heysokam** [reply]
```lua
vim.keymap.set('n', 'K', vim.lsp.buf.hover, { buffer = bufnr, desc = 'vim.lsp.buf.hover()' })
```
I found this in `/usr/share/nvim/runtime/vim/lsp.lua`, but don't know how to disable it from LazyVim :thinking:
Seems like `vim.keymap.del` doesn't want to work from `./config/nvim/lua/config/keymaps.lua` :shrug:

**Reply by heysokam** [reply]
Also tried calling `vim.keymap.del` on `init.lua` right after calling everything else, but it doesn't respond to the deletion either

**Comment by xzbdmw** [comment]
You can just vim.keymap.del("n","K")

**Reply by heysokam** [reply]
> You can just vim.keymap.del("n","K")

Just tried it. Does nothing

**Reply by xzbdmw** [reply]
call it in lspAttach autocmd callback

**Reply by heysokam** [reply]
@xzbdmw
> call it in lspAttach autocmd callback

```lua
    vim.api.nvim_create_autocmd('LspAttach', {
      callback = function(ev)
        vim.bo[ev.buf].formatexpr = nil
        vim.bo[ev.buf].omnifunc = nil
        vim.keymap.del('n', 'K', { buffer = ev.buf })
      end,
    })
    vim.cmd('unmap K')
```
I have this in the `init = function() ...` of `nvim-lspconfig`, like I showed in the `options.lua` file I posted, but it doesn't work
I also added the code snippet (with and without the vim.cmd) to `config/autocmds.lua` and it doesn't work either

In short, I don't know how to do that in LazyVim. What am I missing?

**Reply by xzbdmw** [reply]
That’s unfortunately complex, but I guess what you should try is wrap the call in vim.defer_fn of 1000ms, so after builtin lsp map keys, you make sure delet it

**Reply by xzbdmw** [reply]
It’s better put the lspconfig in plugins dir, not in options.lua

**Reply by heysokam** [reply]
> That’s unfortunately complex, but I guess what you should try is wrap the call in vim.defer_fn of 1000ms, so after builtin lsp map keys, you make sure delet it

It doesn't work either, apparently

**Comment by heysokam** [comment]
![image](https://github.com/LazyVim/LazyVim/assets/7308253/859a7909-0002-4c71-8839-cfc33146d2aa)

THIS fixed it. I had to literally stop LazyVim from touching the option at all
It feels like a bug, but @folke should say if I should file an issue or not.

**Reply by dpetka2001** [reply]
From `:h lsp-defaults`
> |K| is mapped to |vim.lsp.buf.hover()| unless |'keywordprg'| is customized or
  a custom keymap for `K` exists.

When I disable `K` with `keys[#keys + 1] = { "K", false }`, then if I do `verbose nmap K` it shows that it's set up from the Neovim lsp runtime and not LazyVim. So, just set `keys[#keys + 1] = { "K", "" }` (to empty string), so that Neovim also doesn't set it itself. It seems to be a special case with Neovim itself.

**Comment by nguyen-quang-phu** [comment, marked as answer]
i same issues with "gr" mapping and this is my solutions. Hope it can help you.
![image](https://github.com/LazyVim/LazyVim/assets/43996827/f0457bc1-c843-473e-abd2-dca81297e0de)

**Reply by heysokam** [reply]
This worked. Tysm!

**Comment by alainrk** [comment]
I have the same issue now, but with the new nvim 0.11, so without the `neovim/nvim-lspconfig` plugin. Does anyone know what's the best way to do it? I tried through an autocmd but with no success. Thanks all!

**Comment by jonny-rimek** [comment]
this was way harder to fix than it should be, this is in my keymaps.lua

```lua
-- Map J/K for BUFFER navigation (removed K from <nop> section)
vim.keymap.set("n", "J", ":bprevious<CR>", { noremap = true, silent = true, desc = "Previous buffer" })
vim.keymap.set("n", "K", ":bnext<CR>", { noremap = true, silent = true, desc = "Next buffer" })

-- Force K to remain bnext even after LSP attaches
vim.api.nvim_create_autocmd("LspAttach", {
	callback = function(args)
		vim.schedule(function()
			-- Delete LSP's K mapping
			pcall(vim.keymap.del, "n", "K", { buffer = args.buf })
			-- Re-apply our mapping
			vim.keymap.set("n", "K", ":bnext<CR>", {
				buffer = args.buf,
				noremap = true,
				silent = true,
				desc = "Next buffer",
			})
		end)
	end,
})

```

**Reply by dpetka2001** [reply]
If you simply do 
```lua
-- ~/.config/nvim/lua/config/keymaps.lua
vim.keymap.set("n", "K", ":bnext<CR>", { noremap = true, silent = true, desc = "Next buffer" })

-- ~/.config/nvim/lua/plugins/lsp.lua
return {
  "neovim/nvim-lspconfig",
  opts = {
    servers = {
      ["*"] = {
        keys = {
          { "K", false },
        },
      },
    },
  },
}
```
It should work as well.
